scripts/protocol-dts-generator.ts

```diff
@@ -75,153 +66,197 @@ const emitDomain = (domain: P.Domain) => {
     const domainName = toTitleCase(domain.domain)
     emitLine()
     emitDescription(domain.description)
-    emitOpenBlock(`export namespace ${domainName}`)
-    domain.types ? domain.types.forEach(emitType) : null
-    const commandDefs = domain.commands ? domain.commands.map(c => emitCommand(c, domainName)) : []
-    const eventDefs = domain.events ? domain.events.map(e => emitEvent(e, domainName)) : []
+    emitOpenBlock(`export module ${domainName}`)
+    if (domain.types) domain.types.forEach(emitDomainType)
+    if (domain.commands) domain.commands.forEach(emitCommand)
+    if (domain.events) domain.events.forEach(emitEvent)
     emitCloseBlock()
-    const clientDefs = [...commandDefs.map(d => d.client), ...eventDefs.map(d => d.client)]
-    emitInterface(`${domainName}Api`, clientDefs)
-}
-
-const getDomainDef = (domainName: string, $ref:string ) => {
-    return {name: domainName, $ref: `${domainName}${$ref}`}
 }
 
 const getCommentLines = (description: string) => {
     const lines = description
-        .replace(/<code>(.*)<\/code>/g, "'$1'")
         .split(/\r?\n/g)
         .map(line => ` * ${line}`)
-    return [`/**`, ...lines, ,` */`]
+    return [`/**`, ...lines, ` */`]
 }
 
-const emitDescription = (description: string) => {
-    description ? getCommentLines(description).map(l => emitLine(l)) : null
+const emitDescription = (description?: string) => {
+    if (description) getCommentLines(description).map(l => emitLine(l))
 }
 
-const getPropertyDef = (prop: P.ParameterType) => {
-    if ((<any>prop).type == "function") {
-        return `${prop.name}${getPropertyType(prop)}`
-    } else  {
-        return `${prop.name}${prop.optional ? '?' : ''}: ${getPropertyType(prop)}`
-    }
+const getPropertyDef = (prop: P.PropertyType): string => {
+    // Quote key if it has a . in it.
+    const propName = /\./.test(prop.name) ? `'${prop.name}'` : prop.name
+    return `${propName}${prop.optional ? '?' : ''}: ${getPropertyType(prop)}`
 }
 
-const getPropertyType = (prop: any): string  => {
-    if (prop.$ref)
+const getPropertyType = (prop: P.ProtocolType): string  => {
+    if ('$ref' in prop)
         return prop.$ref
-    else if (prop.type == 'array')
+    else if (prop.type === 'array')
         return `${getPropertyType(prop.items)}[]`
-    else if (prop.type == 'object')
-        return `any`
-    else if (prop.type == 'function')
-        return `(${prop.accepts.map(getPropertyDef).join(', ')}): ${prop.returns || 'void'}`
-    else if (prop.type == 'lambda')
-        return `(${prop.accepts.map(getPropertyDef).join(', ')}) => ${prop.returns || 'void'}`
-    else if (prop.type == 'string' && prop.enum)
+    else if (prop.type === 'object')
+        if (!prop.properties) {
+            // TODO: actually 'any'? or can use generic '[key: string]: string'?
+            return `any`
+        } else {
+            // hack: access indent, \n directly
```

Comment:
You could make this into a less hacky version by implementing a `startSubEmit()` and `endSubEmit()`. startSubEmit saves old str and indent in a stack. all emitX calls go to a new str, so you can you use `emitLine()`, `emiOpenBlock` and `emitCloseBlock` without guilt. endSubEmit() would return the str that was generated by the emitCalls.

I know you're only using it here, but I imagine the next dev may follow hack pattern to make more hacks. Up-to-you though, don't wanna bikeshed too much.

Reply:
> You could make this into a less hacky version by implementing a startSubEmit() and endSubEmit()

In the original form of this PR I reworked the whole emit system just for this one spot, but it ended up seeming like way too many changes for what's just a simple reach into the indent state, so I reverted all that and did this instead.

> I know you're only using it here, but I imagine the next dev may follow hack pattern to make more hacks.

Agreed, but maybe the (eventual) need for a second hack will be a better justification for the refactor, and give better insight on what form it should take.